src/bio/bio_device.c

```diff
@@ -456,3 +455,140 @@ bio_replace_dev(struct bio_xs_context *xs_ctxt, uuid_t old_dev_id,
 		smd_free_dev_info(new_info);
 	return rc;
 }
+
+char *
+bio_dev_state_enum_to_str(enum bio_dev_state state)
+{
+	switch (state) {
+	case BIO_DEV_NORMAL: return "NORMAL";
+	case BIO_DEV_FAULTY: return "EVICTED";
+	case BIO_DEV_OUT:    return "UNPLUGGED";
+	case BIO_DEV_NEW:    return "NEW";
+	}
+
+	return "Undefined state";
+}
+
+static struct bio_dev_info *
+alloc_dev_info(uuid_t dev_id, struct smd_dev_info *s_info)
+{
+	struct bio_dev_info	*info;
+	int			 tgt_cnt = 0, i;
+
+	D_ALLOC_PTR(info);
+	if (info == NULL)
+		return NULL;
+
+	if (s_info != NULL) {
+		tgt_cnt = s_info->sdi_tgt_cnt;
+		info->bdi_flags |= NVME_DEV_FL_INUSE;
+		if (s_info->sdi_state == SMD_DEV_FAULTY)
+			info->bdi_flags |= NVME_DEV_FL_FAULTY;
+	}
+
+	if (tgt_cnt != 0) {
+		D_ALLOC_ARRAY(info->bdi_tgts, tgt_cnt);
+		if (info->bdi_tgts == NULL) {
+			D_FREE(info);
+			return NULL;
+		}
+	}
+
+	D_INIT_LIST_HEAD(&info->bdi_link);
+	uuid_copy(info->bdi_dev_id, dev_id);
+	info->bdi_tgt_cnt = tgt_cnt;
+	for (i = 0; i < info->bdi_tgt_cnt; i++)
+		info->bdi_tgts[i] = s_info->sdi_tgts[i];
+
+	return info;
+}
+
+static struct smd_dev_info *
+find_smd_dev(uuid_t dev_id, d_list_t *s_dev_list)
+{
+	struct smd_dev_info	*s_info;
+
+	d_list_for_each_entry(s_info, s_dev_list, sdi_link) {
+		if (uuid_compare(s_info->sdi_id, dev_id) == 0)
+			return s_info;
+	}
+
+	return NULL;
+}
+
+int
+bio_dev_list(struct bio_xs_context *xs_ctxt, d_list_t *dev_list, int *dev_cnt)
+{
+	d_list_t		 s_dev_list;
+	struct bio_dev_info	*b_info;
+	struct smd_dev_info	*s_info, *s_tmp;
+	struct bio_bdev		*d_bdev;
+	int			 rc;
+
+	/* Caller ensures the request handling ULT created on init xstream */
+	D_ASSERT(is_init_xstream(xs_ctxt));
+
+	D_ASSERT(dev_list != NULL && d_list_empty(dev_list));
+	D_INIT_LIST_HEAD(&s_dev_list);
+
+	rc = smd_dev_list(&s_dev_list, dev_cnt);
+	if (rc) {
+		D_ERROR("Failed to get SMD dev list "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	*dev_cnt = 0;
+	/* Scan all plugged devices */
+	d_list_for_each_entry(d_bdev, bio_bdev_list(), bb_link) {
+		s_info = find_smd_dev(d_bdev->bb_uuid, &s_dev_list);
+
+		b_info = alloc_dev_info(d_bdev->bb_uuid, s_info);
+		if (b_info == NULL) {
+			D_ERROR("Failed to allocate device info\n");
+			rc = -DER_NOMEM;
+			goto out;
+		}
+		if (!d_bdev->bb_removed)
+			b_info->bdi_flags |= NVME_DEV_FL_PLUGGED;
+
+		/* set BIO device state based on flags */
```

Comment:
I don't think we should convert the flags into the displayed state here.

These flags are introduced for passing device state internally. With these three flags, we can describe all the device state combinations accurately. (for example, we can tell if an unplugged device was marked as faulty by these flags, but there is no such displayed state) 

The displayed state should only be used by control plane. I don't have strong opinion on how to display device state to end user, but for internal device state description, it should be done by these flags.

Reply:
I understand your argument here, trying to figure out the best place to instead determine the states based on the flags in the control plane. @tanabarr any suggestions?

Reply:
you can pass bdi_flags back in the drpc response and use bitwise operations with imported flag definitions e.g. NVME_DEV_FL_FAULTY to display a state just like you are doing below but in the control-plane server, then just pass strings back to DMG. that makes sense . or did I misunderstand your question?